### PR TITLE
refactor(ast, ast_codegen): use `generate_derive` for implementing `GetSpan` and `GetSpanMut` traits.

### PR DESCRIPTION
**Pull request by rzvxa**

### CODE REVIEW

**Comment by rzvxa** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4735?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4735** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4735?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4732** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4732?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4731** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4731?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4728** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4728?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4730** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4730?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4729** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4729?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4726** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4726?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by rzvxa** [comment]
@Boshen I saw your comment on using `#[generate_derive(GetSpan)]` a tad bit late. Do you want me to refactor it?

File: tasks/ast_codegen/src/generators/derive_get_span.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+    let header = generated_header!();
+
+    quote! {
+        #header
+        insert!("#![allow(clippy::match_same_arms)]");
+        endl!();
+
+        use crate::ast::*;
+        use oxc_span::{#trait_ident, Span};
```
Comment (overlookmotel) [inline tasks/ast_codegen/src/generators/derive_get_span.rs:106]: A nit:

```suggestion
        use oxc_span::{#trait_ident, Span};
        endl!();
        use crate::ast::*;
```

File: tasks/ast_codegen/src/generators/derive_get_span.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    generics: &Option<Generics>,
+    target_type: &Type,
+    self_: &TokenStream,
+    result_type: &TokenStream,
+    body: &TokenStream,
+) -> TokenStream {
+    quote! {
+        endl!();
+        impl #generics #trait_ident for #target_type {
+            #[inline]
```
Comment (overlookmotel) [inline tasks/ast_codegen/src/generators/derive_get_span.rs:149]: Previously `#[inline]` was only added for structs (if I remember right). I think it's unwise to put `#[inline]` on the enums as some of them are pretty large.

File: tasks/ast_codegen/src/generators/derive_get_span.rs (hunk 1 above)
Comment (rzvxa) [reply tasks/ast_codegen/src/generators/derive_get_span.rs]: done

File: tasks/ast_codegen/src/generators/derive_get_span.rs (hunk 2 above)
Comment (rzvxa) [reply tasks/ast_codegen/src/generators/derive_get_span.rs]: Yeah, That was by mistake. Fixed it.

For ast builder I only inline enums with 5 or fewer variants, We can do something similar here as a follow-up.

**Review by overlookmotel — APPROVED** [review]
In my opinion, the code style here is a bit convoluted. But let's merge and I'll make a tidy-up PR later on so we can discuss whether my proposed changes are better or not.

**Pull request merged**